### ISSUE TITLE
Router events runs multiple times.

### ISSUE THREAD

**Issue by boutzamat** [inertiajs/inertia-laravel#583]
When i'm using the Inertia Router start event, all previous events gets logged too.
I'm using the following code: 

```
router.on("start", (event) => {
    console.log(event);

    console.log(`Navigated to ${event}`);

    console.log("--------------");
});
```

This code is running in the main Layout, so it shouldn't be loaded more than once.

The purple box on the image is the console output when i click any link first time, the red box is when i click it the second time (i get both previous event and current event) and if i click it again it will continue like that (recursively).

Is this a bug, or am i doing it wrong?
<img width="724" alt="Skærmbillede 2024-01-24 kl  14 55 17" src="https://github.com/inertiajs/inertia-laravel/assets/4305046/e32ff725-a277-4d19-adc6-7307190b66f6">

**Comment by craigrileyuk** [comment]
How are you calling your layout?

It’s looks like the layout setup is running with each page load, this is to be expected if your layout isn’t persistent.

**Comment by vitalijalbu** [comment]
In DEV mode maybe it's ok because everything runs twice...

**Comment by boutzamat** [comment]
> How are you calling your layout?
> 
> It’s looks like the layout setup is running with each page load, this is to be expected if your layout isn’t persistent.

This is an example of my index page that uses the Layout component:
```
<script setup>
import DeveloperLayout from "@/Layouts/DeveloperLayout.vue";
import PrimaryButton from "@/Components/PrimaryButton.vue";
import { PlusIcon } from "@heroicons/vue/24/outline";
import { Link, usePage } from "@inertiajs/vue3";
import { Head } from "@inertiajs/vue3";

const props = usePage().props;
</script>

<template>
    <DeveloperLayout>
        <Head title="Mine projekter" />

        <template #header>Mine projekter</template>
        <template #header_button>
            <PrimaryButton type="link" :href="route('project.create')">
                <PlusIcon class="mr-2 w-4" />
                Opret projekt
            </PrimaryButton>
        </template>

        <section v-if="!props.projects" class="rounded-3xl bg-white p-16">
            Du har ingen projekter.
            <Link :href="route('project.create')" class="text-primary underline"
                >Opret dit første projekt her </Link
            >.
        </section>
    </DeveloperLayout>
</template>
```

Pretty much the out of the box way of loading the DeveloperLayout. The DeveloperLayout is just the default Layout component renamed.
I've tried with a fresh Laravel-Inertia installation too, and the issue is the same.
The event is triggered in the DeveloperLayout component.

**Comment by boutzamat** [comment]
> In DEV mode maybe it's ok because everything runs twice...

I just tested in production mode. The issue is the same.

**Comment by craigrileyuk** [comment]
Try following the instructions for [Persistent Layouts](https://inertiajs.com/pages#persistent-layouts) instead of wrapping your layout around your page.

The way you've done it, the layout is unmounted and then re-mounted every single time you navigate. This is likely causing the event listener to be registered multiple times.

**Comment by kewcoder** [comment]
i got same issue

**Comment by kewcoder** [comment]
any update? who will fix this issue

**Comment by RobertBoes** [comment]
@kewcoder This isn't an issue here, it's due to how events work and how this is rendered.
It happens because the layout is rerendered, which in turn would bind the listener multiple times, as the listener is never unregistered.

So one way to register and unregister the events would be something along these lines:
```js
import { onMounted, onUnmounted } from 'vue'

let removeEventListener = () => {}

onMounted(() => {
  removeEventListener = router.on("start", (event) => { ... })
}

onUnmounted(() => {
  removeEventListener()
}
```

Inertia also has this approach mentioned in the docs;
```js
onUnmounted(
  router.on('start', (event) => {
    console.log(`Starting a visit to ${event.detail.visit.url}`)
  })
)
```
https://inertiajs.com/events#removing-listeners

So in short, you'd need to be aware of removing event listeners, otherwise more and more listeners would be added.

**Comment by boutzamat** [comment]
It seems to be unfixable as it is how nested components works.
If you want to specify a layout for a specific route/route pattern, you should do it in the app.js file
This way, you don't have to load the layout component per page.

PingCMS is doing layout per page, so i thought this was the recommended way to do it, but it obviously isn't.
This isn't a PingCMS or Inertia issue, but how Vue works.

So, on Dashboard.vue, don't do this:

```
<script>
import Layout from "./Layouts/DefaultLayout.vue";
</script>
<template>
<DefaultLayout>
<div id="dashboard">Hello from Dashboard</div>
</DefaultLayout>
</template>
```


Instead, do this:
**app.js**
```
import Layout from "./Layouts/DefaultLayout.vue";
import SomeOtherLayout from "./Layouts/SomeOtherLayout.vue";

createInertiaApp({
	title: (title) => `${title} - ${appName}`,
	resolve: (name) =>
		resolvePageComponent(`./Pages/${name}.vue`, import.meta.glob("./Pages/**/*.vue")).then(
			(page) => {

				// Default layout for pages
				page.default.layout = Layout;

				// Some other layout logic, i.e. for dashboard
				if (name.startsWith("Dashboard/"))
					page.default.layout = SomeOtherLayout;

				return page;
			}
		),
```

And then in the Dashboard.vue, you don't have to worry about what Layout is being loaded, but just simply do the layout for the dashboard:
```
<template>
<div id="dashboard">Hello from Dashboard</div>
</template>
```

This way, the Layout is only loaded ONCE, and not everytime you change the page.

As @craigrileyuk mentioned in https://github.com/inertiajs/inertia-laravel/issues/583#issuecomment-1987423846, you can also do persistent layouts if you want to control the layout per page.
https://inertiajs.com/pages#persistent-layouts

I hope this helps someone else.

Closing as completed.
